# Pandas

In [3]:
import pandas as pd

# iris = pd.read_csv(Path)

# This function will going to return a file_obj  and there are alot of functions like html reader , text reader , csv reader

In [5]:
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data')
print(type(iris))

<class 'pandas.core.frame.DataFrame'>


# This is refering to iris if we change one of them changes will reflect in both

In [6]:
df = iris

# Here we are creating a copy of iris 

In [7]:
df = iris.copy()

# Head gives us some rows from which we can understand what type of data presend here and t

In [9]:
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa

# we can pass no of rows we want to ge

In [10]:
df.head(3)

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa

# Shape of df 

In [11]:
print(df.shape)

(149, 5)


In [12]:
print(df.dtypes)

5.1            float64
3.5            float64
1.4            float64
0.2            float64
Iris-setosa     object
dtype: object


# Changing column headers

In [13]:
df.columns

Index(['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], dtype='object')

In [14]:
df.columns = ['sl' , 'sw' , 'pl' , 'pw' , 'flower_type']

In [15]:
print(df.dtypes)

sl             float64
sw             float64
pl             float64
pw             float64
flower_type     object
dtype: object


In [34]:
 df.describe()

sl          sw          pl          pw
count  145.000000  145.000000  145.000000  145.000000
mean     5.846897    3.033103    3.848966    1.257931
std      0.914471    0.439548    1.730411    0.783924
min      1.000000    2.000000    1.000000    0.100000
25%      5.100000    2.800000    1.600000    0.400000
50%      5.800000    3.000000    4.400000    1.400000
75%      6.400000    3.300000    5.100000    1.800000
max      7.900000    4.400000    6.900000    4.000000

# Count means how many entries in a column  we have 149 rows so it gives us 149 but some times we can get less then 149 because some of them have NAN values it might be possible 

In [ ]:
# We can acces a particular coulmn 

In [18]:
df.sl

0      4.9
1      4.7
2      4.6
3      5.0
4      5.4
      ... 
144    6.7
145    6.3
146    6.5
147    6.2
148    5.9
Name: sl, Length: 149, dtype: float64

In [19]:
df["sl"]

0      4.9
1      4.7
2      4.6
3      5.0
4      5.4
      ... 
144    6.7
145    6.3
146    6.5
147    6.2
148    5.9
Name: sl, Length: 149, dtype: float64

In [20]:
df.isnull()

sl     sw     pl     pw  flower_type
0    False  False  False  False        False
1    False  False  False  False        False
2    False  False  False  False        False
3    False  False  False  False        False
4    False  False  False  False        False
..     ...    ...    ...    ...          ...
144  False  False  False  False        False
145  False  False  False  False        False
146  False  False  False  False        False
147  False  False  False  False        False
148  False  False  False  False        False

[149 rows x 5 columns]

In [22]:
df.isnull().sum()

sl             0
sw             0
pl             0
pw             0
flower_type    0
dtype: int64

# Slicing 

In [27]:
df.iloc[1:4 , 2:4]
# df[1:4 , 2:4]
# We cant use this 

pl   pw
1  1.3  0.2
2  1.5  0.2
3  1.4  0.2

In [3]:
import pandas as pd
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data')
print(type(iris))

<class 'pandas.core.frame.DataFrame'>


In [4]:
df = iris.copy()

In [6]:
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

# DROP FUNCTION -> This dont change df this will make copy with changes

In [8]:
df.drop(0)
df.head()
# 0th Row is still here so we should take in other variable 

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [9]:
a = df.drop(0)
a.head()
# Now it is here

5.1  3.5  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa

# If we want to change in df itself 

In [11]:
df.drop( 0, inplace = True)
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa

# In drop function drop takes label not index so if once 0 is delete then next time it will give error that 0 is not here

In [12]:
df.columns

Index(['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], dtype='object')

# We can look at labels also 

In [13]:
df.index

RangeIndex(start=1, stop=149, step=1)

In [14]:
df.index[0]
# This will give us the label at position 0 

1

In [ ]:
# If we want to remove by position we use this 

In [15]:
df.drop(df.index[0] , inplace = True)
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa

# We can remove more than One also Here we are giving a range or array 

In [16]:
df.drop(df.index[0],inplace = True)
df.drop(df.index[[0,1]],inplace = True)
df.head()

5.1  3.5  1.4  0.2  Iris-setosa
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa
7  4.4  2.9  1.4  0.2  Iris-setosa
8  4.9  3.1  1.5  0.1  Iris-setosa
9  5.4  3.7  1.5  0.2  Iris-setosa

In [18]:
df.columns = ['sl' , 'sw' , 'pl' , 'pw' , 'flower_type']
df.head()

sl   sw   pl   pw  flower_type
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa
7  4.4  2.9  1.4  0.2  Iris-setosa
8  4.9  3.1  1.5  0.1  Iris-setosa
9  5.4  3.7  1.5  0.2  Iris-setosa

In [ ]:
# We can create an bool_arr 

In [20]:
df.sl > 5

5      False
6      False
7      False
8      False
9       True
       ...  
144     True
145     True
146     True
147     True
148     True
Name: sl, Length: 144, dtype: bool

In [31]:
df[df.sl>5]
x = df[df.sl>5]
print(type(x))
# This will give only that row in which sl is greater than 5 

<class 'pandas.core.frame.DataFrame'>


In [23]:
df[df.flower_type == "Iris-virginica"]

sl   sw   pl   pw     flower_type
99   6.3  3.3  6.0  2.5  Iris-virginica
100  5.8  2.7  5.1  1.9  Iris-virginica
101  7.1  3.0  5.9  2.1  Iris-virginica
102  6.3  2.9  5.6  1.8  Iris-virginica
103  6.5  3.0  5.8  2.2  Iris-virginica
104  7.6  3.0  6.6  2.1  Iris-virginica
105  4.9  2.5  4.5  1.7  Iris-virginica
106  7.3  2.9  6.3  1.8  Iris-virginica
107  6.7  2.5  5.8  1.8  Iris-virginica
108  7.2  3.6  6.1  2.5  Iris-virginica
109  6.5  3.2  5.1  2.0  Iris-virginica
110  6.4  2.7  5.3  1.9  Iris-virginica
111  6.8  3.0  5.5  2.1  Iris-virginica
112  5.7  2.5  5.0  2.0  Iris-virginica
113  5.8  2.8  5.1  2.4  Iris-virginica
114  6.4  3.2  5.3  2.3  Iris-virginica
115  6.5  3.0  5.5  1.8  Iris-virginica
116  7.7  3.8  6.7  2.2  Iris-virginica
117  7.7  2.6  6.9  2.3  Iris-virginica
118  6.0  2.2  5.0  1.5  Iris-virginica
119  6.9  3.2  5.7  2.3  Iris-virginica
120  5.6  2.8  4.9  2.0  Iris-virginica
121  7.7  2.8  6.7  2.0  Iris-virginica
122  6.3  2.7  4.9  1.8  Iris-virginica
123  6.7  3.3  5.7  2.1  Iris-virginica
124  7.2  3.2  6.0  1.8  Iris-virginica
125  6.2  2.8  4.8  1.8  Iris-virginica
126  6.1  3.0  4.9  1.8  Iris-virginica
127  6.4  2.8  5.6  2.1  Iris-virginica
128  7.2  3.0  5.8  1.6  Iris-virginica
129  7.4  2.8  6.1  1.9  Iris-virginica
130  7.9  3.8  6.4  2.0  Iris-virginica
131  6.4  2.8  5.6  2.2  Iris-virginica
132  6.3  2.8  5.1  1.5  Iris-virginica
133  6.1  2.6  5.6  1.4  Iris-virginica
134  7.7  3.0  6.1  2.3  Iris-virginica
135  6.3  3.4  5.6  2.4  Iris-virginica
136  6.4  3.1  5.5  1.8  Iris-virginica
137  6.0  3.0  4.8  1.8  Iris-virginica
138  6.9  3.1  5.4  2.1  Iris-virginica
139  6.7  3.1  5.6  2.4  Iris-virginica
140  6.9  3.1  5.1  2.3  Iris-virginica
141  5.8  2.7  5.1  1.9  Iris-virginica
142  6.8  3.2  5.9  2.3  Iris-virginica
143  6.7  3.3  5.7  2.5  Iris-virginica
144  6.7  3.0  5.2  2.3  Iris-virginica
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  2.0  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica

# Add A row 

In [ ]:
# Add A row 
print(df.head())
print(df.iloc[0])
print(df.loc[0])
# Because there is any 0 label row 

In [28]:
print(df.loc[8])
# It is giving right ans 

sl                     4.9
sw                     3.1
pl                     1.5
pw                     0.1
flower_type    Iris-setosa
Name: 8, dtype: object


# Iloc takes postion/Index and loc takes label 

In [29]:
df.loc[0]= [1,2,3,4,"iris-setosa"]
df.head()
# Its not added to head 

sl   sw   pl   pw  flower_type
5  4.6  3.4  1.4  0.3  Iris-setosa
6  5.0  3.4  1.5  0.2  Iris-setosa
7  4.4  2.9  1.4  0.2  Iris-setosa
8  4.9  3.1  1.5  0.1  Iris-setosa
9  5.4  3.7  1.5  0.2  Iris-setosa

In [30]:
df.tail()

sl   sw   pl   pw     flower_type
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  2.0  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica
0    1.0  2.0  3.0  4.0     iris-setosa

In [ ]:
# Ya its added in end 

In [ ]:
# There is a function value-count fucntion

In [35]:
df.index

Int64Index([  5,   6,   7,   8,   9,  10,  11,  12,  13,  14,
            ...
            140, 141, 142, 143, 144, 145, 146, 147, 148,   0],
           dtype='int64', length=145)

In [ ]:
# Reset Function

In [36]:
df.reset_index()
# This will create a new column 

index   sl   sw   pl   pw     flower_type
0        5  4.6  3.4  1.4  0.3     Iris-setosa
1        6  5.0  3.4  1.5  0.2     Iris-setosa
2        7  4.4  2.9  1.4  0.2     Iris-setosa
3        8  4.9  3.1  1.5  0.1     Iris-setosa
4        9  5.4  3.7  1.5  0.2     Iris-setosa
..     ...  ...  ...  ...  ...             ...
140    145  6.3  2.5  5.0  1.9  Iris-virginica
141    146  6.5  3.0  5.2  2.0  Iris-virginica
142    147  6.2  3.4  5.4  2.3  Iris-virginica
143    148  5.9  3.0  5.1  1.8  Iris-virginica
144      0  1.0  2.0  3.0  4.0     iris-setosa

[145 rows x 6 columns]

In [38]:
df.reset_index(drop = True , inplace = True)
df.index

RangeIndex(start=0, stop=145, step=1)

In [ ]:
# How to add column and delete column

In [ ]:
df.drop('sl')
# This will give an error 

In [40]:
df.drop('sl' , axis= 1 , inplace = True)


KeyError: "['sl'] not found in axis"

In [41]:
df.describe()

sw          pl          pw
count  145.000000  145.000000  145.000000
mean     3.033103    3.848966    1.257931
std      0.439548    1.730411    0.783924
min      2.000000    1.000000    0.100000
25%      2.800000    1.600000    0.400000
50%      3.000000    4.400000    1.400000
75%      3.300000    5.100000    1.800000
max      4.400000    6.900000    4.000000

In [ ]:
# Delete Function 

In [42]:
del df['sw']

In [43]:
df.describe()

pl          pw
count  145.000000  145.000000
mean     3.848966    1.257931
std      1.730411    0.783924
min      1.000000    0.100000
25%      1.600000    0.400000
50%      4.400000    1.400000
75%      5.100000    1.800000
max      6.900000    4.000000

In [45]:
df = iris.copy()
df.columns = ['sl' , 'sw' , 'pl' , 'pw' , 'flower_type']
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [ ]:
# Creating A new Colmun 

In [46]:
df["diff_pl_pw"] = df["pl"] - df["pw"]
df.tail()

sl   sw   pl   pw     flower_type  diff_pl_pw
144  6.7  3.0  5.2  2.3  Iris-virginica         2.9
145  6.3  2.5  5.0  1.9  Iris-virginica         3.1
146  6.5  3.0  5.2  2.0  Iris-virginica         3.2
147  6.2  3.4  5.4  2.3  Iris-virginica         3.1
148  5.9  3.0  5.1  1.8  Iris-virginica         3.3

In [48]:
df["m"] = 1
df.describe()

sl          sw          pl          pw  diff_pl_pw      m
count  149.000000  149.000000  149.000000  149.000000  149.000000  149.0
mean     5.848322    3.051007    3.774497    1.205369    2.569128    1.0
std      0.828594    0.433499    1.759651    0.761292    1.047707    0.0
min      4.300000    2.000000    1.000000    0.100000    0.800000    1.0
25%      5.100000    2.800000    1.600000    0.300000    1.400000    1.0
50%      5.800000    3.000000    4.400000    1.300000    2.900000    1.0
75%      6.400000    3.300000    5.100000    1.800000    3.300000    1.0
max      7.900000    4.400000    6.900000    2.500000    4.700000    1.0

In [49]:
# How to handle NAN 
# 1.) Remove NAN Rows (dropna)
# 2.) Fill NAN (fillna)
df

sl   sw   pl   pw     flower_type  diff_pl_pw  m
0    4.9  3.0  1.4  0.2     Iris-setosa         1.2  1
1    4.7  3.2  1.3  0.2     Iris-setosa         1.1  1
2    4.6  3.1  1.5  0.2     Iris-setosa         1.3  1
3    5.0  3.6  1.4  0.2     Iris-setosa         1.2  1
4    5.4  3.9  1.7  0.4     Iris-setosa         1.3  1
..   ...  ...  ...  ...             ...         ... ..
144  6.7  3.0  5.2  2.3  Iris-virginica         2.9  1
145  6.3  2.5  5.0  1.9  Iris-virginica         3.1  1
146  6.5  3.0  5.2  2.0  Iris-virginica         3.2  1
147  6.2  3.4  5.4  2.3  Iris-virginica         3.1  1
148  5.9  3.0  5.1  1.8  Iris-virginica         3.3  1

[149 rows x 7 columns]

In [ ]:

df.drop('m' , axis=1 , inplace = True)

In [55]:
import numpy as np
df.head()

sl   sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.0  1.4  0.2  Iris-setosa         1.2
1  4.7  3.2  1.3  0.2  Iris-setosa         1.1
2  4.6  0.0  0.0  0.2  Iris-setosa         1.3
3  5.0  0.0  0.0  0.2  Iris-setosa         1.2
4  5.4  3.9  1.7  0.4  Iris-setosa         1.3

In [56]:
df.iloc[2:4 , 1:3] = np.nan
df.head()

sl   sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.0  1.4  0.2  Iris-setosa         1.2
1  4.7  3.2  1.3  0.2  Iris-setosa         1.1
2  4.6  NaN  NaN  0.2  Iris-setosa         1.3
3  5.0  NaN  NaN  0.2  Iris-setosa         1.2
4  5.4  3.9  1.7  0.4  Iris-setosa         1.3

In [ ]:
# We use numpy library which has nan function 

In [57]:
# We can remove nan 
df.dropna(inplace = True)

In [58]:
df.head()

sl   sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.0  1.4  0.2  Iris-setosa         1.2
1  4.7  3.2  1.3  0.2  Iris-setosa         1.1
4  5.4  3.9  1.7  0.4  Iris-setosa         1.3
5  4.6  3.4  1.4  0.3  Iris-setosa         1.1
6  5.0  3.4  1.5  0.2  Iris-setosa         1.3

In [60]:
# Now we can see both row delete
df.reset_index(drop= True , inplace = True)

In [61]:
df.head()

sl   sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.0  1.4  0.2  Iris-setosa         1.2
1  4.7  3.2  1.3  0.2  Iris-setosa         1.1
2  5.4  3.9  1.7  0.4  Iris-setosa         1.3
3  4.6  3.4  1.4  0.3  Iris-setosa         1.1
4  5.0  3.4  1.5  0.2  Iris-setosa         1.3

In [67]:
# We can fill some values in nan 
df.iloc[2:4 , 1:3] = np.nan
df.head()

sl   sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.0  1.4  0.2  Iris-setosa         1.2
1  4.7  3.2  1.3  0.2  Iris-setosa         1.1
2  5.4  NaN  NaN  0.4  Iris-setosa         1.3
3  4.6  NaN  NaN  0.3  Iris-setosa         1.1
4  5.0  3.4  1.5  0.2  Iris-setosa         1.3

In [68]:
df.sw.fillna(df.sw.mean() , inplace = True)

In [69]:
df.head()

sl        sw   pl   pw  flower_type  diff_pl_pw
0  4.9  3.000000  1.4  0.2  Iris-setosa         1.2
1  4.7  3.200000  1.3  0.2  Iris-setosa         1.1
2  5.4  3.038621  NaN  0.4  Iris-setosa         1.3
3  4.6  3.038621  NaN  0.3  Iris-setosa         1.1
4  5.0  3.400000  1.5  0.2  Iris-setosa         1.3

In [70]:
df.pl.fillna(df.pl.mean() , inplace = True)

In [71]:
df.head()

sl        sw        pl   pw  flower_type  diff_pl_pw
0  4.9  3.000000  1.400000  0.2  Iris-setosa         1.2
1  4.7  3.200000  1.300000  0.2  Iris-setosa         1.1
2  5.4  3.038621  3.837241  0.4  Iris-setosa         1.3
3  4.6  3.038621  3.837241  0.3  Iris-setosa         1.1
4  5.0  3.400000  1.500000  0.2  Iris-setosa         1.3

In [72]:
df["Gender"] = "Female"
df.iloc[0:10 , 6] = "Male"
df.tail()

sl   sw   pl   pw     flower_type  diff_pl_pw  Gender
142  6.7  3.0  5.2  2.3  Iris-virginica         2.9  Female
143  6.3  2.5  5.0  1.9  Iris-virginica         3.1  Female
144  6.5  3.0  5.2  2.0  Iris-virginica         3.2  Female
145  6.2  3.4  5.4  2.3  Iris-virginica         3.1  Female
146  5.9  3.0  5.1  1.8  Iris-virginica         3.3  Female

In [ ]:
df.Gender.apply(f)


In [8]:
import pandas as pd
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data')
df = iris.copy()
print(df.head())

   5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa


In [9]:
df.columns = ["sl", "sw" , "pl" , "pw" , "flower_type"]
df.head()

sl   sw   pl   pw  flower_type
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [15]:
x = df.sl.iloc[::-1]
print(type(x))
print(x)

<class 'pandas.core.series.Series'>
148    5.9
147    6.2
146    6.5
145    6.3
144    6.7
      ... 
4      5.4
3      5.0
2      4.6
1      4.7
0      4.9
Name: sl, Length: 149, dtype: float64


In [4]:
print("hii")

hii


In [11]:
df[df.sl == x] 

ValueError: Can only compare identically-labeled Series objects